### PR TITLE
udt to udt connections

### LINKED ISSUE #6688: User-Defined Resource Types
## Overview of feature request

Today, there are a fixed set of portable Radius resources:

- Applications.Dapr/stateStores
- Applications.Dapr/pubSubBrokers
- Applications.Datastores/redisCaches
- Applications.Datastores/sqlDatabases
- ...

These resources have a fixed schema and it allows teams to only select from a pre-defined list.

What's missing is a way for a team at an organization like Contoso to model their own schema and abstractions to work with.

For example, Contoso may want a resource type like:

```bicep
resource storage 'Contoso.Storage/database' = {
  name: 'inventory'
  properties: {...}
}
```

This would allow developers to select from resource types that are custom to Contoso. This also doesn't have to be limited to storage/messaging resources. Custom types could also encapsulate entire services or dependencies such as:

```bicep
resource storage 'Contoso.Monitoring/diagnostics' = {
  name: 'inventory'
  properties: {...}
}
```

## Acceptance criteria

- Users can define their own types + schema
- These resources can be modeled in Bicep, acting as a key/value store for the properties
- Recipes should work with the resources, allowing IT operators to define Recipe templates that spin up the actual backing infrastructure.

## Additional context

[AB#10284](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/10284)
```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by nithyatsu**
# Description

This PR adds support for UDt to UDT connections. 

Design https://github.com/radius-project/design-notes/pull/81 refer ## Scenario 3 – Connections
## Type of change

- This pull request adds or changes features of Radius and has an approved issue (issue link required).

Partially Fixes: #6688

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

- An overview of proposed schema changes is included in a linked GitHub issue.
    - [X] Yes 
    - [ ] Not applicable 
- A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
    - [X] Yes 
    - [ ] Not applicable 
- The design document has been reviewed and approved by Radius maintainers/approvers.
    - [X] Yes 
    - [ ] Not applicable 
- A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
    - [X] Yes 
    - [ ] Not applicable 
- A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
    - [X] Yes 
    - [ ] Not applicable 
- A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.
    - [ ] Yes 
    - [X] Not applicable

### CODE REVIEW

**Comment by nithyatsu** [comment]
https://github.com/radius-project/samples/issues/2320

https://github.com/radius-project/docs/issues/1435

File: pkg/resourceutil/utils.go [hunk 1]
```diff
@@ -55,3 +57,38 @@ func GetPropertiesFromResource[P any](resource P) (map[string]any, error) {
...
+// It returns a map where the keys are connection names and the values are the corresponding connected resource's IDs.
+func GetConnectionNameandSourceIDs[P any](resource P) (map[string]string, error) {
+	connectionNamesAndSourceIDs := map[string]string{}
+	resourceProperties, err := GetPropertiesFromResource(resource)
+	if err != nil {
+		return nil, err
+	}
+
+	if resourceProperties != nil && resourceProperties["connections"] != nil {
+		connections, ok := resourceProperties["connections"].(map[string]interface{})
```
Comment (lakshmimsft) [inline pkg/resourceutil/utils.go]: pls update to use `any` instead of interface{} in this function and tests

File: pkg/resourceutil/utils_test.go [hunk 2]
```diff
@@ -136,3 +136,224 @@ func TestGetPropertiesFromResource_MissingProperties(t *testing.T) {
...
+	}
+
+	result, err := GetConnectionNameandSourceIDs(resource)
+	require.Error(t, err)
+	require.Nil(t, result)
+	require.Contains(t, err.Error(), errMarshalResource)
+}
+
+func TestGetConnectionNameandSourceIDs_EdgeCases(t *testing.T) {
+	t.Run("Connection with additional properties", func(t *testing.T) {
```
Comment (lakshmimsft) [inline pkg/resourceutil/utils_test.go]: it feels like tests in this function are not really edge cases. resource having other properties is handled in TestGetConnectionNameandSourceIDs() and doesn't feel liek edge cases. Let me know if I'm missing something.

File: test/functional-portable/dynamicrp/noncloud/resources/dynamicrp_test.go [hunk 3]
```diff
@@ -313,3 +315,297 @@ func Test_DynamicRP_ExternalResource(t *testing.T) {
...
+					require.Equal(t, env.Name, env.ValueFrom.SecretKeyRef.Key)
+				}
+
+			},
+		},
+	})
+	test.Test(t)
+}
+
+func Test_UDT_ConnectionTo_UDT(t *testing.T) {
```
Comment (lakshmimsft) [inline test/functional-portable/dynamicrp/noncloud/resources/dynamicrp_test.go]: Can we add documentation for this test similar to above test.

File: test/functional-portable/dynamicrp/noncloud/resources/dynamicrp_test.go [hunk 4]
```diff
@@ -313,3 +315,297 @@ func Test_DynamicRP_ExternalResource(t *testing.T) {
...
+
+				require.True(t, found, "No deployments found with label 'radapp.io/connected-to-resource' in pod template labels")
+
+			},
+		},
+	})
+	test.Test(t)
+}
+
+func Test_UDT_ConnectionTo_UDTTF(t *testing.T) {
```
Comment (lakshmimsft) [inline test/functional-portable/dynamicrp/noncloud/resources/dynamicrp_test.go]: Same as above. Comments needed.

File: test/functional-portable/dynamicrp/noncloud/resources/dynamicrp_test.go [hunk 5]
```diff
@@ -313,3 +315,297 @@ func Test_DynamicRP_ExternalResource(t *testing.T) {
...
+	existingTemplate := "testdata/udt2udt-connection-tf.bicep"
+	name := "dynamicrp-udt2udt"
+	appNamespace := "udttoudtapp"
+	appName := "udttoudtapp"
```
Comment (lakshmimsft) [inline test/functional-portable/dynamicrp/noncloud/resources/dynamicrp_test.go]: Can we have separate appName (and any other resource names in test) for uniqueness from above bicep test.

File: test/functional-portable/dynamicrp/noncloud/resources/testdata/container2udt-connection.bicep [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+param registry string
+
+param version string
+
+@description('PostgreSQL password')
+@secure()
+param password string = newGuid()
+
+resource udtconnenv 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dynamicrp-postgres-env'
```
Comment (lakshmimsft) [inline test/functional-portable/dynamicrp/noncloud/resources/testdata/container2udt-connection.bicep]: resource name 'dynamicrp-postgres-env' is in use in existing test ../test/functional-portable/dynamicrp/noncloud/resources/testdata/postgres-env-scoped-resource.bicep
Pls review resource names used and update to unique names in tests. It will help with debugging in tests and when we move back to running them in parallel.

File: pkg/portableresources/backend/controller/createorupdateresource.go [hunk 7]
```diff
@@ -173,13 +173,35 @@ func (c *CreateOrUpdateResource[P, T]) executeRecipeIfNeeded(ctx context.Context
 		return nil, err
 	}
 
+	connectionsAndSourceIDs, err := resourceutil.GetConnectionNameandSourceIDs(resource)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get connected resource IDs: %w", err)
+	}
+	connectedResourcesProperties := make(map[string]map[string]any)
+	// If there are connected resources, we need to fetch their properties and add them to the recipe context.
```
Comment (lakshmimsft) [inline pkg/portableresources/backend/controller/createorupdateresource.go:182]: nit: add line before comment

File: pkg/portableresources/backend/controller/createorupdateresource.go [hunk 8]
```diff
@@ -173,13 +173,35 @@ func (c *CreateOrUpdateResource[P, T]) executeRecipeIfNeeded(ctx context.Context
...
+	for connName, connectedResourceID := range connectionsAndSourceIDs {
+		connectedResource, err := c.DatabaseClient().Get(ctx, connectedResourceID)
+		if errors.Is(&database.ErrNotFound{ID: connectedResourceID}, err) {
+			return nil, fmt.Errorf("connected resource %s not found: %w", connectedResourceID, err)
+		} else if err != nil {
+			return nil, fmt.Errorf("failed to get connected resource %s: %w", connectedResourceID, err)
+		}
+		connectedResourceProperties, err := resourceutil.GetPropertiesFromResource(connectedResource.Data)
```
Comment (lakshmimsft) [inline pkg/portableresources/backend/controller/createorupdateresource.go:191]: nit: consider add line

File: pkg/resourceutil/utils_test.go (hunk 2 above)
Comment (nithyatsu) [reply pkg/resourceutil/utils_test.go]: Agreed. Removed the claude generated test case.

File: test/functional-portable/dynamicrp/noncloud/resources/testdata/udt2udt-connection.bicep [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+      {
+        kind: 'kubernetesNamespace'
+        namespace: 'udttoudtapp'
+      }
+    ]
+  }
+}
+
+
+resource udttoudtparent 'Test.Resources/udtParent@2023-10-01-preview' = {
```
Comment (lakshmimsft) [inline test/functional-portable/dynamicrp/noncloud/resources/testdata/udt2udt-connection.bicep]: One concern with the test is both resources are heavy postgres resources with request of 512MB each which is going to slow test pieplines down/can cause resource issues. Can we consider other options. eg. Parent UDT is a wrapper around configmap, Child UDT is a k8s pod that uses data from parent and sets it as a label /env variable. 
(sample ref: config map was used in test test/functional-portable/dynamicrp/noncloud/resources/testdata/externalresource.bicep)
This also applies to container-udt test..

File: test/functional-portable/dynamicrp/noncloud/resources/testdata/container2udt-connection.bicep [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    properties: {
+      application: udtapp.id
+      container: {
+        image: 'ghcr.io/radius-project/samples/demo:latest'
+        ports: {
+          web: {
+            containerPort: 3000
+          }
+        }
+
```
Comment (lakshmimsft) [inline test/functional-portable/dynamicrp/noncloud/resources/testdata/container2udt-connection.bicep]: nit: extra lines

**Review by lakshmimsft — APPROVED** [review]
